chainpoint/chainpoint-core
# boltwall integration

This PR removes redis tracking of invoices related to hash submissions and instead uses boltwall to manage endpoint authorization. 

New endpoints include:
- `GET /boltwall/node` to get information about the node that will be paid (useful for establishing better connection to facilitate payment)
- `POST /boltwall/hodl` to create a new hodl invoice. No body is necessary as a middleware will check for or create (if none is found) a `paymentHash` and attach it to the req.body such that boltwall can generate a new invoice from it
- `GET /invoice` If an invoice has been generated in association with the request (determined by macaroons attached as cookies) this will get information about the invoice.

More comments about the code and new functionality will be included inline in the PR.

This PR also includes:
- some additional configs in the docker-compose to enable boltwall's ln-service connection as well as the ability to test against a simnet node running locally (this needs to be made more flexible, open to suggestions!)
- A middleware that creates a unique sessionID associated with new or expired requests. This is used to manage inflight invoices.

**TESTING DIRECTIONS**
1. Make sure you have boltbox available locally
2. run `make simnet` to start up a local simnet network
3. Get alice's `adminMacaroon`, `cert`, and socket info. The first two should be in a local `credentials.json` file generated by the script. The socket info should be `alice:10001`
4. Add the following env vars to your `.env` file
```
LND_SOCKET=alice:10001
LND_TLS_CERT=[alices tls cert]
LND_MACAROON=[alice's admin macaroon]
CAVEAT_KEY=[random string used as a password to sign macaroon caveats]
# used to sign, verify, and decrypt sessions
SESSION_SECRET=[randomly generated 32 byte hex string. Can be left out and generated automatically, but session cookies can't be persisted between server restarts if skipped]
```
(Note: if you want to test against a running testnet node you can replace the relevant connection information above. Please note, that this must be a different node than the one you will be paying from for testing)
5. To get the services up and running first run `docker-compose up -d` from the chainpoint-core directory.
6. Once everything is started, you can stop the api container `docker-compose stop api` and then restart it with the logs (useful for testing) with `docker-compose up --build api`
7. The update with nodemon and the mapped volumes mean that if you make a change (or just save) the server.js or any file in the endpoints directory, it will get picked up in the container and the server restarted. **Useful for testing, but open to feedback on this.**

**Other tools helpful for testing**
I recommend testing with postman (so you can easily parse and add headers) and the [LSAT Playground](https://lsat-playground.bucko.now.sh/). The LSAT playground will allow you to drop in the challenges to generate an LSAT token from it when making your requests. 

**TESTING ENDPOINTS**
I find it easiest to use Postman for this, but whatever works for you is obviously fine
1. First try to `GET /localhost/boltwall/node` to see your node's information
1. Next try and `POST localhost/hash` without any payment processing (include a proper `hash` in the request body). You should get a 402 error.
1. Retrieve the LSAT from the `WWW-Authenticate` header. This will include an invoice. 
1. Use the playground to decode the challenge and get the LSAT token (which is just the macaroon). 
1. Add the token to your request headers under the `Authorization` header.
1. Without paying anything, try and submit a hash again. You should get a response that says the invoice is unpaid and give you the invoice information
1. If your simnet started up correctly, you can head on over to `localhost:5000` and log in to RTL (password should be `foobar)
1. make sure you're on bob or carol's dashboard (check top left of the dashboard) since alice is getting paid. Go to payments > send and enter your payment information and click "Send Payment". This will hang since it's a hodl invoice.
1. Try and `POST localhost/hash` again. Now it should go through and the RTL dashboard should show the invoice as confirmed (no more hanging)
1. Try and submit again, you'll get a 402 response again. 
1. Rinse and repeat!
